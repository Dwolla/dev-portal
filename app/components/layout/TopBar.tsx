import { useContext } from "react";
import Button from "@mui/material/Button";
import styled from "@emotion/styled";
import Link from "next/link";
import { FormControl } from "@mui/material";
import { LanguageContext } from "../util/Contexts";
import { breakDown, breakUp } from "../breakpoints";
import { GREY_2, GREY_4, PURPLE_PRIMARY } from "../colors";
import dwollaDevLogo from "../../../assets/images/logo-developers.png";
import { ReactComponent as CloseIcon } from "../../../assets/images/component-icons/close.svg";
import ga from "../../modules/ga";
import SelectMui from "../base/SelectMui";

export const TOP_BAR_HEIGHT = 66;

const Container = styled.div`
  display: flex;
  align-items: center;
  height: ${TOP_BAR_HEIGHT}px;
  background-color: ${PURPLE_PRIMARY};
  padding: 0 20px;

  @media (${breakDown("md")}) {
    padding: 12px 20px;
  }
`;

const StyledLogo = styled.img`
  width: 15rem;

  @media (${breakDown("xs")}) {
    max-width: 13rem;
  }
`;

const ButtonWrapper = styled.div`
  margin-left: auto;
  padding-right: 20px;

  @media (${breakDown("md")}) {
    display: none;
  }
`;

const SelectWrapper = styled.div`
  padding-left: 20px;

  @media (${breakDown("xs")}) {
    display: none;
  }
`;

export type TopBarButtonProps = {
  text: string;
  link: {
    href: string;
    external: boolean;
  };
};

export type TopBarProps = {
  button: TopBarButtonProps;
  sidebarToggled?: boolean;
  setSidebarToggled?: Function;
};

const Hamburger = styled.div`
  width: 15px;
  height: 12px;
  margin-left: auto;
  cursor: pointer;
  background-image: linear-gradient(
    to bottom,
    ${GREY_2},
    ${GREY_2} 16.67%,
    transparent 16.67%,
    transparent 41.67%,
    ${GREY_2} 41.67%,
    ${GREY_2} 58.33%,
    transparent 58.33%,
    transparent 83.33%,
    ${GREY_2} 83.33%,
    ${GREY_2} 100%
  );

  @media (${breakUp("lg")}) {
    display: none;
  }

  :focus {
    outline: none;
    opacity: 0.5;
  }

  &:hover,
  &.isFocused {
    opacity: 0.5;
  }
`;

const StyledCloseIcon = styled.div`
  cursor: pointer;
  margin-left: auto;
  svg {
    * {
      fill: ${GREY_2};
    }
  }

  :hover {
    svg {
      * {
        fill: ${GREY_4};
      }
    }
  }

  :focus {
    outline: none;
    svg {
      * {
        fill: ${GREY_4};
      }
    }
  }

  @media (${breakUp("lg")}) {
    display: none;
  }
`;

export default function TopBar({
  button,
  sidebarToggled,
  setSidebarToggled,
}: TopBarProps) {
  const { selectedLanguage, setSelectedLanguage, languageOptions } =
    useContext(LanguageContext);

  const showSidebar = () => setSidebarToggled(true);
  const hideSidebar = () => setSidebarToggled(false);

  function onHamburgerKeydown(e) {
    if (e.key === "Enter") showSidebar();
  }

  function onCloseIconKeydown(e) {
    if (e.key === "Enter") hideSidebar();
  }

  return (
    <Container>
      <Link href="/">
        <a>
          <StyledLogo src={dwollaDevLogo} alt="" />
        </a>
      </Link>

      <SelectWrapper>
        <FormControl sx={{ minWidth: 200 }} size="small">
          <SelectMui
            color="white"
            label="Select Language"
            onChange={(value) => {
              setSelectedLanguage(value);
              ga("global language select", "language change", value);
            }}
            options={languageOptions}
            value={selectedLanguage.value}
          />
        </FormControl>
      </SelectWrapper>

      <ButtonWrapper>
        <Button
          variant="contained"
          href={button.link.href}
          target={button.link.external ? "_blank" : undefined}
          color="secondary"
          size="medium"
        >
          {button.text}
        </Button>
      </ButtonWrapper>

      {sidebarToggled === false ? (
        <Hamburger
          tabIndex={0}
          aria-labelledby="Open menu"
          onClick={showSidebar}
          onKeyDown={(keyPress) => onHamburgerKeydown(keyPress)}
          // onFocus={() => setIsFocused(true)}
          // onBlur={() => setIsFocused(false)}
        />
      ) : (
        <StyledCloseIcon
          tabIndex={0}
          aria-labelledby="Close menu"
          onClick={hideSidebar}
          onKeyDown={(keyPress) => onCloseIconKeydown(keyPress)}
          // onFocus={() => setIsFocused(true)}
          // onBlur={() => setIsFocused(false)}
        >
          <CloseIcon width={14} />
        </StyledCloseIcon>
      )}
    </Container>
  );
}
