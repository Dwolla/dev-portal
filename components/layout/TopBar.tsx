import { useContext, useState } from "react";
import styled from "@emotion/styled";
import classnames from "classnames";
import Link from "next/link";
import { useRouter } from "next/router";
import openInNewTabIcon from "../../assets/images/component-icons/open-in-new-tab-icon.svg";
import { GREY_2, GREY_6, ORANGE_PRIMARY } from "../colors";
import { POPPINS } from "../typography";
import Button from "../base/Button";
import Select from "../base/select/Select";
import { LanguageContext } from "../util/Contexts";
import { breakUp, breakDown } from "../breakpoints";
import dwollaDevLogo from "../../assets/images/dwolla-developers-logo.svg";
import ga from "../../modules/ga";

export const TOP_BAR_HEIGHT = 68;

const Container = styled.div`
  display: flex;
  align-items: center;
  height: ${TOP_BAR_HEIGHT}px;
  border-bottom: 1px solid ${GREY_2};
  background: rgba(255, 255, 255, 0.98);
  padding: 0 20px;

  @media (${breakDown("md")}) {
    padding: 12px 20px;
  }
`;

const StyledLogoWrap = styled.div`
  @media (${breakUp("lg")}) {
    display: none;
  }
`;

const StyledLogo = styled.img`
  max-height: 100%;

  @media (${breakDown("xs")}) {
    max-width: 115px;
  }
`;

const LinksWrapper = styled.div`
  height: 100%;

  @media (${breakDown("md")}) {
    display: none;
  }
`;

const ButtonWrapper = styled.div`
  margin-left: auto;
  padding-right: 20px;

  @media (${breakDown("md")}) {
    display: none;
  }
`;

const SelectWrapper = styled.div`
  padding-left: 20px;
`;

const StyledLink = styled.a`
  font-family: ${POPPINS};
  font-size: 14px;
  color: ${GREY_6};
  display: inline-flex;
  align-items: center;
  padding: 0 20px;
  height: 100%;
  text-decoration: none;

  > .text {
    display: inline-flex;
    align-items: center;
    height: 100%;
    border-bottom: 3px solid transparent;
  }

  > .icon {
    margin-left: 5px;
    width: 13px;
    height: 13px;
  }

  &.active > .text {
    color: ${ORANGE_PRIMARY};
    border-bottom-color: ${ORANGE_PRIMARY};
  }

  :hover:not(.active) > .text {
    text-decoration: underline;
  }
`;

export type TopBarLinkProps = {
  href: string;
  external?: boolean;
  text: string;
  // manually set a link as active as opposed to relying on useRouter().pathname
  active?: boolean;
};

const ConditionalLinkWrapper = ({ external, linkWrapper, children }) =>
  external ? children : linkWrapper(children);

export function TopBarLink({ href, external, text, active }: TopBarLinkProps) {
  return (
    <ConditionalLinkWrapper
      external={external}
      linkWrapper={(children) => (
        <Link href={href} passHref>
          {children}
        </Link>
      )}
    >
      <StyledLink
        href={external ? href : ""}
        target={external && "_blank"}
        className={classnames({ active })}
      >
        <span className="text">{text}</span>

        {external && (
          <img className="icon" src={openInNewTabIcon} alt="Open in new tab" />
        )}
      </StyledLink>
    </ConditionalLinkWrapper>
  );
}

export type TopBarButtonProps = {
  text: string;
  link: {
    href: string;
    external: boolean;
  };
};

export type TopBarProps = {
  button: TopBarButtonProps;
  links: TopBarLinkProps[];
  onHamburgerClick?: () => void;
};

const Hamburger = styled.div`
  width: 15px;
  height: 12px;
  margin-left: auto;
  cursor: pointer;
  background-image: linear-gradient(
    to bottom,
    #000,
    #000 16.67%,
    transparent 16.67%,
    transparent 41.67%,
    #000 41.67%,
    #000 58.33%,
    transparent 58.33%,
    transparent 83.33%,
    #000 83.33%,
    #000 100%
  );

  @media (${breakUp("lg")}) {
    display: none;
  }

  :focus {
    outline: none;
  }

  &:hover,
  &.isFocused {
    opacity: 0.5;
  }
`;

export default function TopBar({
  button,
  links,
  onHamburgerClick,
}: TopBarProps) {
  const { selectedLanguage, setSelectedLanguage, languageOptions } = useContext(
    LanguageContext
  );

  const [isFocused, setIsFocused] = useState(false);

  function onHamburgerKeydown(e) {
    if (e.key === "Enter") onHamburgerClick();
  }
  const router = useRouter();

  return (
    <Container>
      <StyledLogoWrap>
        <Link href="/">
          <a>
            <StyledLogo src={dwollaDevLogo} alt="" />
          </a>
        </Link>
      </StyledLogoWrap>

      <LinksWrapper>
        {links.map((l) => (
          <TopBarLink
            key={l.href}
            {...l}
            active={l.active || l.href === router?.pathname}
          />
        ))}
      </LinksWrapper>

      <SelectWrapper>
        <Select
          options={languageOptions}
          selectedValue={selectedLanguage}
          setSelectedValue={(language) => {
            setSelectedLanguage(language);
            ga("global language select", "language change", language.value);
          }}
          autoWidth
        />
      </SelectWrapper>

      <ButtonWrapper>
        <Button {...button} size="standard" variant="primary" />
      </ButtonWrapper>

      {onHamburgerClick && (
        <Hamburger
          tabIndex={0}
          aria-labelledby="Open menu"
          className={classnames({ isFocused })}
          onClick={onHamburgerClick}
          onKeyDown={(keyPress) => onHamburgerKeydown(keyPress)}
          onFocus={() => setIsFocused(true)}
          onBlur={() => setIsFocused(false)}
        />
      )}
    </Container>
  );
}
