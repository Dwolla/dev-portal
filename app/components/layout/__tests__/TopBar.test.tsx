import renderer from "react-test-renderer";
import TopBar from "../TopBar";

const mockPath = "/changelog";
jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: mockPath,
      pathname: mockPath,
      query: "",
      asPath: mockPath,
    };
  },
}));

test("TopBar", () => {
  const tree = renderer
    .create(
      <TopBar
        button={{
          text: "Get API Keys",
          link: {
            href: "https://accounts-sandbox.dwolla.com/login",
            external: true,
          },
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
