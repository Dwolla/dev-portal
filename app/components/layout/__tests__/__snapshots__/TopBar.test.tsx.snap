// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar 1`] = `
<div
  className="css-16rtae5-Container e1psldpd0"
>
  <div
    className="css-hafdn5-StyledLogoWrap e1psldpd1"
  >
    <a
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <img
        alt=""
        className="css-g6dikn-StyledLogo e1psldpd2"
        src="test-image-stub"
      />
    </a>
  </div>
  <div
    className="css-ffwcst-LinksWrapper e1psldpd3"
  >
    <a
      className=" css-skf8jj-StyledLink e1psldpd6"
      href="https://docs.dwolla.com"
      target="_blank"
    >
      <span
        className="text"
      >
        API Docs
      </span>
      <img
        alt="Open in new tab"
        className="icon"
        src="test-image-stub"
      />
    </a>
    <a
      className="active css-skf8jj-StyledLink e1psldpd6"
      href="/changelog"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <span
        className="text"
      >
        Changelog
      </span>
    </a>
    <a
      className="active css-skf8jj-StyledLink e1psldpd6"
      href="/manually-active"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <span
        className="text"
      >
        Manually Active
      </span>
    </a>
  </div>
  <div
    className="css-zhv85h-SelectWrapper e1psldpd5"
  >
    <div
      className=" css-k377xg-container"
      onKeyDown={[Function]}
    >
      <div
        className=" css-1qsurod-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className=" css-15ikjri-ValueContainer"
        >
          <div
            className=" css-1me8bqy-singleValue"
          >
            
          </div>
          <input
            aria-autocomplete="list"
            className="css-62g3xt-dummyInput"
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            readOnly={true}
            tabIndex="0"
            value=""
          />
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className=" css-43ykx9-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className=" css-unjupu-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <img
              alt=""
              className=" css-vfym8x-StyledDefaultDropdownIcon evf7wlc0"
              src="test-image-stub"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-1as8qu1-ButtonWrapper e1psldpd4"
  >
    <a
      className="primary standard  css-5fh7ca-StyledLink-buttonStyles-buttonStyles ek3ainu1"
      href="https://accounts-sandbox.dwolla.com/login"
      target="_blank"
    >
      Get API Keys
    </a>
  </div>
</div>
`;
